### DISCUSSION TITLE
Best practices and antipatterns

### DISCUSSION THREAD

**Discussion by diberry** [devcontainers/community#107] [Q&A]
I've been handed several sample projects which use dev containers to ease customer adoption. The samples, because they are complex, may take 20 or 30 minutes to fully open on a local computer. Generally speaking, this time is take when the dev container uses a postCreateCommand in a monorepo (several packages/containers/crates/apps) to install each package and set up any data ingestion such as database migration into the docker-in-docker database service. 

I understand the purpose for this but I think it is front-loading that time/installation in a way that makes the dev container _appear_ like a terrible dev experience. 

Are there best practices where I can go back to the repo owners and guide them to a better dev experience with some clear dos and donts?

**Comment by bamurtaugh** [comment]
Thanks for sharing this @diberry. 

A fantastic first step would be guiding folks to prebuild images as much as possible: https://containers.dev/guide/prebuild. 

If you have any questions on the prebuild process, and/or would like to share any specific dev container configs for some feedback on them, I'm more than happy to discuss further.
